flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/JoinUtil.scala

```diff
@@ -0,0 +1,385 @@
+      cnfCondition: RexNode,
+      leftFieldCount: Int,
+      inputType: RelDataType): (Seq[(RelDataType, Boolean, RexNode)], Seq[RexNode]) = {
+
+    cnfCondition match {
+      case c: RexCall if c.getKind == SqlKind.AND =>
+        val timeIndicators =
+          c.operands.map(splitJoinCondition(_, leftFieldCount, inputType))
+        timeIndicators.reduceLeft { (l, r) =>
+          (l._1 ++ r._1, l._2 ++ r._2)
+        }
+      case c: RexCall =>
+        val timeIndicators = analyzeSingleConditionTerm(c, leftFieldCount, inputType)
+        timeIndicators match {
+          case Seq() =>
+            (Seq(), Seq(c))
+          case Seq(v1, v2) =>
+            if (v1._1 != v2._1) {
+              throw new TableException("The time indicators for each input should be the same.")
+            }
+            if (v1._2 == v2._2) {
+              throw new TableException("Both input's time indicators is needed.")
+            }
+            (Seq((v1._1, v1._2, c)), Seq())
+          case _ =>
+            throw new TableException(
+              "There only can and must have one time indicators for each input.")
+        }
+      case other =>
+        val timeIndicators = analyzeSingleConditionTerm(other, leftFieldCount, inputType)
+        timeIndicators match {
+          case Seq() =>
+            (Seq(), Seq(other))
+          case _ =>
+            throw new TableException("Time indicators can not be used in non time-condition.")
+        }
+    }
+  }
+
+  /**
+   * analysis if condition term has time indicator
+   */
+  def analyzeSingleConditionTerm(
+      expression: RexNode,
+      leftFieldCount: Int,
+      inputType: RelDataType): Seq[(RelDataType, Boolean)] = {
+
+    expression match {
+      case i: RexInputRef =>
+        val idx = i.getIndex
+        inputType.getFieldList.get(idx).getType match {
+          case t: TimeIndicatorRelDataType if idx < leftFieldCount =>
+            // left table time indicator
+            Seq((t, true))
+          case t: TimeIndicatorRelDataType =>
+            // right table time indicator
+            Seq((t, false))
+          case _ => Seq()
+        }
+      case c: RexCall =>
+        c.operands.map(analyzeSingleConditionTerm(_, leftFieldCount, inputType)).reduce(_++_)
+      case _ => Seq()
+    }
+  }
+
+  /**
+    * Extract time offset and determain which table the offset belong to
+    */
+  def extractTimeOffsetFromCondition(
+      timeTerm: RexNode,
+      isLeftExprBelongLeftTable: Boolean,
+      rexBuilder: RexBuilder,
+      config: TableConfig) = {
+
+    val timeCall: RexCall = timeTerm.asInstanceOf[RexCall]
+    val leftLiteral =
+      reduceTimeExpression(
+        timeCall.operands.get(0),
+        rexBuilder,
+        config)
+
+    val rightLiteral =
+      reduceTimeExpression(
+        timeCall.operands.get(1),
+        rexBuilder,
+        config)
+
+    val (tmpTimeOffset: Long, isLeftTableTimeOffset: Boolean) =
+      timeTerm.getKind match {
```

Comment:
for the case which both in the same side like ```a.proctime - b.proctime > 5 sec```, it will report an error ```Cannot apply '-' to arguments of type '<TIMESTAMP(3)> - <TIMESTAMP(3)> '```. So i think it will not exist such situation.